src/ComptimeInterpreter.zig

```diff
@@ -0,0 +1,1454 @@
+//! Hacky comptime interpreter, courtesy of midnight code run fuelled by spite;
+//! hope that one day this can use async... <33
+
+// TODO: DODify
+
+const std = @import("std");
+const ast = @import("ast.zig");
+const zig = std.zig;
+const Ast = zig.Ast;
+const analysis = @import("analysis.zig");
+const DocumentStore = @import("DocumentStore.zig");
+const ComptimeInterpreter = @This();
+
+const log = std.log.scoped(.comptime_interpreter);
+
+// TODO: Investigate arena
+
+allocator: std.mem.Allocator,
+document_store: *DocumentStore,
+handle: *const DocumentStore.Handle,
+root_type: ?Type = null,
+
+/// Interpreter diagnostic errors
+errors: std.AutoArrayHashMapUnmanaged(Ast.Node.Index, InterpreterError) = .{},
+
+// TODO: Deduplicate typeinfo across different interpreters
+type_info: std.ArrayListUnmanaged(TypeInfo) = .{},
+type_info_map: std.HashMapUnmanaged(TypeInfo, usize, TypeInfo.Context, std.hash_map.default_max_load_percentage) = .{},
+
+// TODO: Use DOD
+value_data_list: std.ArrayListUnmanaged(*ValueData) = .{},
+
+pub const InterpreterError = struct {
+    code: []const u8,
+    message: []const u8,
+};
+
+/// `message` must be allocated with interpreter allocator
+pub fn recordError(interpreter: *ComptimeInterpreter, node_idx: Ast.Node.Index, code: []const u8, message: []const u8) error{OutOfMemory}!void {
+    try interpreter.errors.put(interpreter.allocator, node_idx, .{
+        .code = code,
+        .message = message,
+    });
+}
+
+pub fn deinit(interpreter: *ComptimeInterpreter) void {
+    var err_it = interpreter.errors.iterator();
+    while (err_it.next()) |entry| interpreter.allocator.free(entry.value_ptr.message);
+
+    if (interpreter.root_type) |rt| rt.getTypeInfo().getScopeOfType().?.deinit();
+    for (interpreter.type_info.items) |*ti| ti.deinit(interpreter.allocator);
+    for (interpreter.value_data_list.items) |ti| interpreter.allocator.destroy(ti);
+
+    interpreter.errors.deinit(interpreter.allocator);
+    interpreter.type_info.deinit(interpreter.allocator);
+    interpreter.type_info_map.deinit(interpreter.allocator);
+    interpreter.value_data_list.deinit(interpreter.allocator);
+}
+
+pub const TypeInfo = union(enum) {
+    pub const Context = struct {
+        interpreter: ComptimeInterpreter,
+        hasher: *std.hash.Wyhash,
+
+        pub fn hash(self: @This(), s: TypeInfo) u64 {
+            TypeInfo.hash(self, s);
+            return self.hasher.final();
+        }
+        pub fn eql(self: @This(), a: TypeInfo, b: TypeInfo) bool {
+            _ = self;
+            return TypeInfo.eql(a, b);
+        }
+    };
+
+    pub const Signedness = enum { signed, unsigned };
+
+    pub const Struct = struct {
+        /// Declarations contained within
+        scope: *InterpreterScope,
+        fields: std.StringHashMapUnmanaged(FieldDefinition) = .{},
+    };
+
+    pub const Int = struct {
+        bits: u16,
+        signedness: Signedness,
+    };
```

Comment:
```suggestion
    pub const Struct = struct {
        /// Declarations contained within
        scope: *InterpreterScope,
        fields: std.StringHashMapUnmanaged(FieldDefinition) = .{},
    };

    pub const Int = struct {
        bits: u16,
        signedness: std.builtin.Signedness,
    };
```
you can use `std.builtin.Signedness` here